File: src/2025-02-23-deepclaude-try/java/exercises/practice/rest-api/src/main/java/RestApi.java
import org.json.JSONObject;

import java.util.*;
import org.json.JSONObject;
import org.json.JSONArray;

class RestApi {
    private final Map<String, User> users = new HashMap<>();

    RestApi(User... initialUsers) {
        for (User user : initialUsers) {
            users.put(user.name(), user);
        }
    }

    String get(String url) {
        if ("/users".equals(url)) {
            return getUsersResponse(new ArrayList<>(users.values()));
        }
        throw new IllegalArgumentException("Unknown URL: " + url);
    }

    String get(String url, JSONObject payload) {
        if ("/users".equals(url) && payload.has("users")) {
            JSONArray userNames = payload.getJSONArray("users");
            List<User> requestedUsers = new ArrayList<>();
            for (int i = 0; i < userNames.length(); i++) {
                String name = userNames.getString(i);
                if (users.containsKey(name)) {
                    requestedUsers.add(users.get(name));
                }
            }
            return getUsersResponse(requestedUsers);
        }
        throw new IllegalArgumentException("Invalid request");
    }

    String post(String url, JSONObject payload) {
        if ("/add".equals(url)) {
            String name = payload.getString("user");
            if (users.containsKey(name)) {
                return createUserJson(users.get(name)).toString();
            }
            User newUser = User.builder().setName(name).build();
            users.put(name, newUser);
            return createUserJson(newUser).toString();
        }
        
        if ("/iou".equals(url)) {
            String lender = payload.getString("lender");
            String borrower = payload.getString("borrower");
            double amount = payload.getDouble("amount");
            
            User lenderUser = users.get(lender);
            User borrowerUser = users.get(borrower);
            
            User.Builder lenderBuilder = User.builder().setName(lender);
            User.Builder borrowerBuilder = User.builder().setName(borrower);
            
            // Copy existing IOUs
            for (Iou iou : lenderUser.owes()) {
                lenderBuilder.owes(iou.name(), iou.amount());
            }
            for (Iou iou : lenderUser.owedBy()) {
                lenderBuilder.owedBy(iou.name(), iou.amount());
            }
            for (Iou iou : borrowerUser.owes()) {
                borrowerBuilder.owes(iou.name(), iou.amount());
            }
            for (Iou iou : borrowerUser.owedBy()) {
                borrowerBuilder.owedBy(iou.name(), iou.amount());
            }
            
            // Add new IOU
            lenderBuilder.owedBy(borrower, amount);
            borrowerBuilder.owes(lender, amount);
            
            User updatedLender = lenderBuilder.build();
            User updatedBorrower = borrowerBuilder.build();
            
            users.put(lender, updatedLender);
            users.put(borrower, updatedBorrower);
            
            List<User> updatedUsers = Arrays.asList(updatedLender, updatedBorrower);
            Collections.sort(updatedUsers, Comparator.comparing(User::name));
            return getUsersResponse(updatedUsers);
        }
        
        throw new IllegalArgumentException("Unknown URL: " + url);
    }

    private String getUsersResponse(List<User> userList) {
        JSONObject response = new JSONObject();
        JSONArray usersArray = new JSONArray();
        
        Collections.sort(userList, Comparator.comparing(User::name));
        
        for (User user : userList) {
            usersArray.put(createUserJson(user));
        }
        
        response.put("users", usersArray);
        return response.toString();
    }

    private JSONObject createUserJson(User user) {
        JSONObject userJson = new JSONObject();
        userJson.put("name", user.name());
        
        JSONObject owes = new JSONObject();
        for (Iou iou : user.owes()) {
            owes.put(iou.name(), iou.amount());
        }
        userJson.put("owes", owes);
        
        JSONObject owedBy = new JSONObject();
        for (Iou iou : user.owedBy()) {
            owedBy.put(iou.name(), iou.amount());
        }
        userJson.put("owed_by", owedBy);
        
        double balance = user.owedBy().stream().mapToDouble(Iou::amount).sum() -
                        user.owes().stream().mapToDouble(Iou::amount).sum();
        userJson.put("balance", balance);
        
        return userJson;
    }
}

File: src/python-2025-02-27-deepclaude37-rust/python/exercises/practice/rest-api/rest_api.py
class RestAPI:
    def __init__(self, database=None):
        self.database = database if database is not None else {'users': []}

    def get(self, url, payload=None):
        import json
        if url == '/users':
            requested_users = []
            if payload:
                payload_data = json.loads(payload)
                requested_users = payload_data.get('users', [])
            users = self.database['users']
            if requested_users:
                filtered_users = [u for u in users if u['name'] in requested_users]
            else:
                filtered_users = users.copy()
            sorted_users = sorted(filtered_users, key=lambda x: x['name'])
            response = []
            for user in sorted_users:
                owed = sum(user['owed_by'].values())
                owes = sum(user['owes'].values())
                response.append({
                    'name': user['name'],
                    'owes': user['owes'],
                    'owed_by': user['owed_by'],
                    'balance': owed - owes
                })
            return json.dumps({'users': response})
        return json.dumps({})

    def post(self, url, payload=None):
        import json
        if url == '/add':
            data = json.loads(payload)
            new_user = {
                'name': data['user'],
                'owes': {},
                'owed_by': {}
            }
            self.database['users'].append(new_user)
            return json.dumps({
                'name': new_user['name'],
                'owes': new_user['owes'],
                'owed_by': new_user['owed_by'],
                'balance': 0.0
            })
        elif url == '/iou':
            data = json.loads(payload)
            lender_name = data['lender']
            borrower_name = data['borrower']
            amount = data['amount']
            
            lender = next(u for u in self.database['users'] if u['name'] == lender_name)
            borrower = next(u for u in self.database['users'] if u['name'] == borrower_name)
            
            # Process IOU - handle existing debts between users
            if lender_name in borrower['owed_by']:
                owed = borrower['owed_by'][lender_name]
                if owed >= amount:
                    borrower['owed_by'][lender_name] -= amount
                    if borrower['owed_by'][lender_name] == 0:
                        del borrower['owed_by'][lender_name]
                    lender['owes'][borrower_name] -= amount
                    if lender['owes'][borrower_name] == 0:
                        del lender['owes'][borrower_name]
                else:
                    remaining = amount - owed
                    del borrower['owed_by'][lender_name]
                    if borrower_name in lender['owes']:
                        del lender['owes'][borrower_name]
                    borrower['owes'][lender_name] = borrower['owes'].get(lender_name, 0) + remaining
                    lender['owed_by'][borrower_name] = lender['owed_by'].get(borrower_name, 0) + remaining
            else:
                borrower['owes'][lender_name] = borrower['owes'].get(lender_name, 0) + amount
                lender['owed_by'][borrower_name] = lender['owed_by'].get(borrower_name, 0) + amount
            
            # Prepare response with updated user data
            users = [lender, borrower]
            users.sort(key=lambda x: x['name'])
            response = []
            for user in users:
                owed = sum(user['owed_by'].values())
                owes = sum(user['owes'].values())
                response.append({
                    'name': user['name'],
                    'owes': user['owes'],
                    'owed_by': user['owed_by'],
                    'balance': owed - owes
                })
            return json.dumps({'users': response})
        return json.dumps({})
